### PR TITLE
Add integration tests for apps and groups 

### PR DESCRIPTION
**Pull request by aquamatthias**
...that execute the marathon testplan for releases.

@kolloch Please Review

### CODE REVIEW

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala [hunk 1]
```diff
@@ -19,9 +23,256 @@ class AppDeployIntegrationTest
...
+    val app = appProxy(testBasePath / "http-app", "v1", instances = 1, withHealth = false).
+      copy(healthChecks = Set(HealthCheck(gracePeriod = 20.second, interval = 1.second, maxConsecutiveFailures = 10)))
+
+    When("The app is deployed")
+    val result = marathon.createApp(app)
+
+    Then("The app is created")
+    result.code should be (201) //Created
+    extractDeploymentIds(result) should have size 1
+    waitForEvent("deployment_success")
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: In the integration test plan it says that one should check that the health checks actually have been started.

Maybe one could check the count of healthyTasks from the app definition?

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala [hunk 2]
```diff
@@ -19,9 +23,256 @@ class AppDeployIntegrationTest
...
+  }
+
+  test("create a simple app with command health checks") {
+    Given("a new app")
+    val app = appProxy(testBasePath / "command-app", "v1", instances = 1, withHealth = false).
+      copy(healthChecks = Set(HealthCheck(protocol = Protocol.COMMAND, command = Some(Command("true")), gracePeriod = 20.second, interval = 1.second, maxConsecutiveFailures = 10)))
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: Quite a long line. It's actually slightly annoying reviewing it like this.

But if you disagree heavily, I do not care that much.

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala [hunk 3]
```diff
@@ -19,9 +23,256 @@ class AppDeployIntegrationTest
...
+
+    When("the deployment has finished")
+    waitForEvent("deployment_success")
+
+    Then("the list of running apps can be fetched")
+    val apps = marathon.listAppsInBaseGroup
+    apps.code should be(200)
+    apps.value should have size 1
+
+    val tasks = marathon.tasks(appId)
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: I head to go to my IDE to check the method here. Maybe its nice to have a type annotation here.

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala [hunk 4]
```diff
@@ -19,9 +23,256 @@ class AppDeployIntegrationTest
...
+    tasks.code should be(200)
+    tasks.value should have size 2
+  }
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: the test plan is more specific (with a sample result)

I guess we can deal with that later though by recording/comparing the API interactions

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala (hunk 3 above)
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: Then("the list of running app tasks can be fetched")

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala [hunk 5]
```diff
@@ -19,9 +23,256 @@ class AppDeployIntegrationTest
...
+    Given("a new app")
+    val app = appProxy(testBasePath / "app", "v1", instances = 1, withHealth = false)
+
+    When("The app is deployed")
+    val result = marathon.createApp(app)
+
+    Then("The app is created")
+    result.code should be (201) //Created
+    extractDeploymentIds(result) should have size 1
+    waitForEvent("deployment_success")
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: should we maybe check that the app is really running?

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala [hunk 6]
```diff
@@ -19,9 +23,256 @@ class AppDeployIntegrationTest
...
+    val before = marathon.tasks(appId)
+
+    When("The app is updated")
+    val update = marathon.updateApp(v1.id, AppUpdate(cmd = appProxy(appId, "v2", 1).cmd))
+
+    Then("The app gets updated")
+    update.code should be (200)
+    waitForEvent("deployment_success")
+    val after = marathon.tasks(appId)
+    waitForTasks(appId, before.value.size)
```
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: should we check that the tasks conform to the new definition?

**Comment by kolloch** [comment]
Nice! Apart from the comments I wonder if we can speed this up, though. For example by other settings in mesos or Marathon?

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala (hunk 5 above)
Comment (aquamatthias) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: That would test mesos not marathon - but I can add a test for this.

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala (hunk 2 above)
Comment (aquamatthias) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: Change it to a shorter version.

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala (hunk 6 above)
Comment (aquamatthias) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: Added a check, that makes sure the new version is launched.

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala (hunk 5 above)
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: I guess the question is what the test scope is ;) It is an integration test...

Alternatively, you could query the mesos api and check if the task is marked as running or something like that.

But you wouldn't for example test the generation of environment variables like $PORT which is also pretty important.

**Comment by kolloch** [comment]
It would be great if you squashed the commit so that our project history only contains one commit.

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala (hunk 1 above)
Comment (aquamatthias) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: See the test - it is checked, that the app is started.

Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: yes, and it does not check whether the status fields are set correctly. But admittedly that's a matter of defining the test scope.

File: src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala (hunk 6 above)
Comment (kolloch) [inline src/test/scala/mesosphere/marathon/integration/AppDeployIntegrationTest.scala]: You do not check that, do you?

You could check it by inspecting `after.value.map(_.version)` for example.

**Pull request merged**